src/neurostatslib/glm.py

```diff
@@ -1,431 +1,598 @@
-        if (predicted_firing_rates == 0).any():
-            warnings.warn(
-                "predicted_firing_rates array contained zeros, this can "
-                "lead to infinite log-likelihood values."
-            )
-        window_size = self.spike_basis_matrix.shape[1]
-        return self._score(predicted_firing_rates, spike_data[:, window_size:])
+        # check if the model is fit
+        self._check_is_fit()
+        Ws, bs = self.basis_coeff_, self.baseline_link_fr_
+        (feedforward_input,) = self.preprocess_simulate(
+            feedforward_input, params_f=(Ws, bs)
+        )
+        predicted_rate = self._predict((Ws, bs), feedforward_input)
+        return (
+            self.noise_model.emission_probability(
+                key=random_key, predicted_rate=predicted_rate
+            ),
+            predicted_rate,
+        )
 
-    def simulate(
+
+class GLMRecurrent(GLM):
+    """
+    A Generalized Linear Model (GLM) with recurrent dynamics.
+
+    This class extends the basic GLM to capture recurrent dynamics between neurons,
+    making it more suitable for simulating the activity of interconnected neural populations.
+    The recurrent GLM combines both feedforward inputs (like sensory stimuli) and past
+    neural activity to simulate or predict future neural activity.
+
+    Parameters
+    ----------
+    noise_model : nsm.NoiseModel, default=nsm.PoissonNoiseModel()
+        The noise model to use for the GLM. This defines how neural activity is generated
+        based on the underlying firing rate. Common choices include Poisson and Gaussian models.
+
+    solver : slv.Solver, default=slv.RidgeSolver()
+        The optimization solver to use for fitting the GLM parameters.
+
+    data_type : {jnp.float32, jnp.float64}, optional
+        The numerical data type for internal calculations. If not provided, it will be inferred
+        from the data during fitting.
+
+    See Also
+    --------
+    [GLM](../glm/#neurostatslib.glm.GLM) : Base class for the generalized linear model.
+
+    Notes
+    -----
+    - The recurrent GLM assumes that neural activity can be influenced by both feedforward
+    inputs and the past activity of the same and other neurons. This makes it particularly
+    powerful for capturing the dynamics of neural networks where neurons are interconnected.
+
+    - The attributes of `GLMRecurrent` are inherited from the parent `GLM` class, and might include
+    coefficients, fitted status, and other model-related attributes.
+
+    Examples
+    --------
+    >>> # Initialize the recurrent GLM with default parameters
+    >>> model = GLMRecurrent()
+    >>> # ... your code for training and simulating using the model ...
+
+    """
+
+    def __init__(
+        self,
+        noise_model: nsm.NoiseModel = nsm.PoissonNoiseModel(),
+        solver: slv.Solver = slv.RidgeSolver(),
+    ):
+        super().__init__(noise_model=noise_model, solver=solver)
+
+    def simulate_recurrent(
```

Comment:
this means we have both `simulate` and `simulate_recurrent` for this class -- is that what we want?

Reply:
unsure, what do you think? 
simulate in GLMRecurrent would be simulating the spike times conditioned to the actual activity, while simulate_recurrent would be the free-running simulation (conditioning to the actual generated spikes). I can see how this may create confusion